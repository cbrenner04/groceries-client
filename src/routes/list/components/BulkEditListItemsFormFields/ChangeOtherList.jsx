import React from 'react';
import PropTypes from 'prop-types';
import { Button, Form } from 'react-bootstrap';

import { CheckboxField, SelectField, TextField } from '../../../../components/FormFields';
import { prettyListType } from '../../../../utils/format';

const ChangeOtherList = (props) => {
  const handleMove = () => {
    props.handleOtherListChange(false);
  };

  const handleCopy = () => {
    props.handleOtherListChange(true);
  };

  const changeListInstructions = props.existingListsOptions.length
    ? 'Choose an existing list or create a new one.'
    : `You do not have any other ${prettyListType(props.listType)}s. Please create a new list to take this action.`;

  return (
    <>
      <Form.Group className="form-check-inline mb-3">
        <Form.Check
          name="copy"
          label="Copy"
          type="radio"
          onChange={handleCopy}
          onClick={handleCopy}
          value={props.copy}
          id="move-action-copy"
          checked={props.copy}
        />
        <Form.Check
          name="move"
          label="Move"
          type="radio"
          value={props.move}
          id="move-action-move"
          onChange={handleMove}
          onClick={handleMove}
          checked={props.move}
        />
      </Form.Group>
      {(props.copy || props.move) && (
        <>
          <div className="mb-3">{changeListInstructions}</div>
          {props.showNewListForm && props.existingListsOptions.length > 0 && (
            <Button variant="link" onClick={props.clearNewListForm} className="mb-3">
              Choose existing list
            </Button>
          )}
          {!props.showNewListForm && (
            <>
              <Button variant="link" onClick={props.handleShowNewListForm} className="mb-3">
                Create new list
              </Button>
              <SelectField
                name="existingList"
                label="Existing list"
                value={props.existingList}
                options={props.existingListsOptions}
                handleChange={props.handleInput}
                blankOption
              />
            </>
          )}
          {props.showNewListForm && (
            <TextField
              name="newListName"
              label="New list name"
              value={props.newListName}
              handleChange={props.handleInput}
              placeholder="My super cool list"
            />
          )}
          {props.copy && !props.allComplete && (
            <CheckboxField
              name="updateCurrentItems"
              label="Would you like to also update the current items?"
              handleChange={props.handleInput}
              value={props.updateCurrentItems}
              classes="mt-3"
            />
          )}
        </>
      )}
    </>
  );
};

ChangeOtherList.propTypes = {
  handleOtherListChange: PropTypes.func.isRequired,
  copy: PropTypes.bool.isRequired,
  move: PropTypes.bool.isRequired,
  showNewListForm: PropTypes.bool.isRequired,
  existingListsOptions: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string,
    }),
  ).isRequired,
  listType: PropTypes.string.isRequired,
  handleInput: PropTypes.func.isRequired,
  handleShowNewListForm: PropTypes.func.isRequired,
  clearNewListForm: PropTypes.func.isRequired,
  existingList: PropTypes.string.isRequired,
  newListName: PropTypes.string.isRequired,
  updateCurrentItems: PropTypes.bool.isRequired,
  allComplete: PropTypes.bool.isRequired,
};

export default ChangeOtherList;
