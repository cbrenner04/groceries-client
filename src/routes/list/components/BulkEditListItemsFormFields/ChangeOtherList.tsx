import React, { type ChangeEventHandler } from 'react';
import { Button, Form } from 'react-bootstrap';

import { CheckboxField, SelectField, TextField } from 'components/FormFields';
import { prettyListType } from 'utils/format';

export interface IChangeOtherListProps {
  handleOtherListChange: (isCopy: boolean) => void;
  copy: boolean;
  move: boolean;
  showNewListForm: boolean;
  existingListsOptions: {
    value: string;
    label: string;
  }[];
  listType: string;
  handleInput: ChangeEventHandler;
  handleShowNewListForm: () => void;
  clearNewListForm: () => void;
  existingList: string;
  newListName: string;
  updateCurrentItems: boolean;
  allComplete: boolean;
}

const ChangeOtherList: React.FC<IChangeOtherListProps> = (props): React.JSX.Element => {
  const handleMove = (): void => {
    props.handleOtherListChange(false);
  };

  const handleCopy = (): void => {
    props.handleOtherListChange(true);
  };

  const changeListInstructions = props.existingListsOptions.length
    ? 'Choose an existing list or create a new one.'
    : `You do not have any other ${prettyListType(props.listType)}s. Please create a new list to take this action.`;

  return (
    <React.Fragment>
      <Form.Group className="form-check-inline mb-3">
        <Form.Check
          name="copy"
          label="Copy"
          type="radio"
          onChange={handleCopy}
          onClick={handleCopy}
          value={String(props.copy)}
          id="move-action-copy"
          checked={props.copy}
        />
        <Form.Check
          name="move"
          label="Move"
          type="radio"
          value={String(props.move)}
          id="move-action-move"
          onChange={handleMove}
          onClick={handleMove}
          checked={props.move}
        />
      </Form.Group>
      {(props.copy || props.move) && (
        <React.Fragment>
          <div className="mb-3">{changeListInstructions}</div>
          {props.showNewListForm && props.existingListsOptions.length > 0 && (
            <Button variant="link" onClick={props.clearNewListForm} className="mb-3">
              Choose existing list
            </Button>
          )}
          {!props.showNewListForm && (
            <React.Fragment>
              <Button variant="link" onClick={props.handleShowNewListForm} className="mb-3">
                Create new list
              </Button>
              <SelectField
                name="existingList"
                label="Existing list"
                value={props.existingList}
                options={props.existingListsOptions}
                handleChange={props.handleInput}
                blankOption
              />
            </React.Fragment>
          )}
          {props.showNewListForm && (
            <TextField
              name="newListName"
              label="New list name"
              value={props.newListName}
              handleChange={props.handleInput}
              placeholder="My super cool list"
            />
          )}
          {props.copy && !props.allComplete && (
            <CheckboxField
              name="updateCurrentItems"
              label="Would you like to also update the current items?"
              handleChange={props.handleInput}
              value={props.updateCurrentItems}
              classes="mt-3"
            />
          )}
        </React.Fragment>
      )}
    </React.Fragment>
  );
};

export default ChangeOtherList;
