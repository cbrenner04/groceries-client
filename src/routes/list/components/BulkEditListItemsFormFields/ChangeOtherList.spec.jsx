import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import ChangeOtherList from './ChangeOtherList';

describe('ChangeOtherList', () => {
  let props;

  beforeEach(() => {
    props = {
      handleOtherListChange: jest.fn(),
      copy: false,
      move: false,
      showNewListForm: false,
      existingListsOptions: [
        {
          value: '1',
          label: 'foobar',
        },
      ],
      listType: 'GroceryList',
      handleInput: jest.fn(),
      handleShowNewListForm: jest.fn(),
      clearNewListForm: jest.fn(),
      existingList: '',
      newListName: '',
      updateCurrentItems: false,
    };
  });

  it('renders appropriate change list instructions when existingListsOptions', () => {
    props.copy = true;
    props.existingListsOptions = [
      {
        value: '1',
        label: 'foobar',
      },
    ];
    const { getByText, queryByText } = render(<ChangeOtherList {...props} />);

    expect(getByText('Choose an existing list or create a new one.')).toBeTruthy();
    expect(
      queryByText('You do not have any other Grocery Lists. Please create a new list to take this action.'),
    ).toBeNull();
  });

  it('renders appropriate change list instructions when no existingListsOptions', () => {
    props.copy = true;
    props.existingListsOptions = [];
    const { getByText, queryByText } = render(<ChangeOtherList {...props} />);

    expect(queryByText('Choose an existing list or create a new one.')).toBeNull();
    expect(
      getByText('You do not have any other Grocery Lists. Please create a new list to take this action.'),
    ).toBeTruthy();
  });

  it('renders when copy and handle change when selected', () => {
    props.copy = true;
    const { container, getByLabelText } = render(<ChangeOtherList {...props} />);

    expect(container).toMatchSnapshot();

    fireEvent.click(getByLabelText('Copy'));

    expect(props.handleOtherListChange).toHaveBeenCalledWith(true);
  });

  it('renders when move and handle change when selected', () => {
    props.move = true;
    const { container, getByLabelText } = render(<ChangeOtherList {...props} />);

    expect(container).toMatchSnapshot();

    fireEvent.click(getByLabelText('Move'));

    expect(props.handleOtherListChange).toHaveBeenCalledWith(false);
  });

  it('renders only radios when not copy and not move', () => {
    const { container, getByLabelText, queryByLabelText } = render(<ChangeOtherList {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Copy')).toBeVisible();
    expect(getByLabelText('Move')).toBeVisible();
    expect(queryByLabelText('Existing list')).toBeNull();
    expect(queryByLabelText('New list name')).toBeNull();
    expect(queryByLabelText('Would you like to also update the current items?')).toBeNull();
  });

  it('shows choose existing list link when existingListsOptions and correct fields when showNewList and copy', () => {
    props.copy = true;
    props.showNewListForm = true;
    const { container, getByLabelText, queryByLabelText, getByRole } = render(<ChangeOtherList {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Copy')).toBeVisible();
    expect(getByLabelText('Move')).toBeVisible();
    expect(getByRole('button')).toHaveTextContent('Choose existing list');
    expect(queryByLabelText('Existing list')).toBeNull();
    expect(getByLabelText('New list name')).toBeVisible();
    expect(getByLabelText('Would you like to also update the current items?')).toBeVisible();
  });

  it('shows choose existing list link when existingListsOptions and correct fields when showNewList and move', () => {
    props.move = true;
    props.showNewListForm = true;
    const { container, getByLabelText, queryByLabelText, getByRole } = render(<ChangeOtherList {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Copy')).toBeVisible();
    expect(getByLabelText('Move')).toBeVisible();
    expect(getByRole('button')).toHaveTextContent('Choose existing list');
    expect(queryByLabelText('Existing list')).toBeNull();
    expect(getByLabelText('New list name')).toBeVisible();
    expect(queryByLabelText('Would you like to also update the current items?')).toBeNull();
  });

  it('does not show choose existing list link when not existingListsOptions and showNewList', () => {
    props.move = true;
    props.showNewListForm = true;
    props.existingListsOptions = [];
    const { container, getByLabelText, queryByLabelText, queryByRole } = render(<ChangeOtherList {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Copy')).toBeVisible();
    expect(getByLabelText('Move')).toBeVisible();
    expect(queryByRole('button')).toBeNull();
    expect(queryByLabelText('Existing list')).toBeNull();
    expect(getByLabelText('New list name')).toBeVisible();
    expect(queryByLabelText('Would you like to also update the current items?')).toBeNull();
  });

  it('shows create new list link when not showNewListForm', () => {
    props.copy = true;
    props.showNewListForm = false;
    const { container, getByLabelText, queryByLabelText, getByRole } = render(<ChangeOtherList {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Copy')).toBeVisible();
    expect(getByLabelText('Move')).toBeVisible();
    expect(getByRole('button')).toHaveTextContent('Create new list');
    expect(getByLabelText('Existing list')).toBeVisible();
    expect(queryByLabelText('New list name')).toBeNull();
    expect(getByLabelText('Would you like to also update the current items?')).toBeVisible();
  });
});
