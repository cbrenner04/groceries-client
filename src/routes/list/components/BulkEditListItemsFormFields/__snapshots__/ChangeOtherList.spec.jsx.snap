// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeOtherList does not render update current items when all items are complete 1`] = `
<div>
  <div
    class="form-check-inline mb-3"
  >
    <div
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="move-action-copy"
        name="copy"
        type="radio"
        value="true"
      />
      <label
        class="form-check-label"
        for="move-action-copy"
        title=""
      >
        Copy
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-move"
        name="move"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-move"
        title=""
      >
        Move
      </label>
    </div>
  </div>
  <div
    class="mb-3"
  >
    Choose an existing list or create a new one.
  </div>
  <button
    class="mb-3 btn btn-link"
    type="button"
  >
    Create new list
  </button>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="existingList"
    >
      Existing list
    </label>
    <select
      class="form-control"
      id="existingList"
      name="existingList"
    >
      <option
        disabled=""
        value=""
      >
        Select Existing list
      </option>
      <option
        value="1"
      >
        foobar
      </option>
    </select>
  </div>
</div>
`;

exports[`ChangeOtherList does not show choose existing list link when not existingListsOptions and showNewList 1`] = `
<div>
  <div
    class="form-check-inline mb-3"
  >
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-copy"
        name="copy"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-copy"
        title=""
      >
        Copy
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="move-action-move"
        name="move"
        type="radio"
        value="true"
      />
      <label
        class="form-check-label"
        for="move-action-move"
        title=""
      >
        Move
      </label>
    </div>
  </div>
  <div
    class="mb-3"
  >
    You do not have any other  Grocery Lists. Please create a new list to take this action.
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="newListName"
    >
      New list name
    </label>
    <input
      class="form-control"
      id="newListName"
      name="newListName"
      placeholder="My super cool list"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`ChangeOtherList renders only radios when not copy and not move 1`] = `
<div>
  <div
    class="form-check-inline mb-3"
  >
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-copy"
        name="copy"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-copy"
        title=""
      >
        Copy
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-move"
        name="move"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-move"
        title=""
      >
        Move
      </label>
    </div>
  </div>
</div>
`;

exports[`ChangeOtherList renders when copy and handle change when selected 1`] = `
<div>
  <div
    class="form-check-inline mb-3"
  >
    <div
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="move-action-copy"
        name="copy"
        type="radio"
        value="true"
      />
      <label
        class="form-check-label"
        for="move-action-copy"
        title=""
      >
        Copy
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-move"
        name="move"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-move"
        title=""
      >
        Move
      </label>
    </div>
  </div>
  <div
    class="mb-3"
  >
    Choose an existing list or create a new one.
  </div>
  <button
    class="mb-3 btn btn-link"
    type="button"
  >
    Create new list
  </button>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="existingList"
    >
      Existing list
    </label>
    <select
      class="form-control"
      id="existingList"
      name="existingList"
    >
      <option
        disabled=""
        value=""
      >
        Select Existing list
      </option>
      <option
        value="1"
      >
        foobar
      </option>
    </select>
  </div>
  <div
    class="mt-3 mb-3"
  >
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="updateCurrentItems"
        name="updateCurrentItems"
        type="checkbox"
      />
      <label
        class="form-check-label"
        for="updateCurrentItems"
        title=""
      >
        Would you like to also update the current items?
      </label>
    </div>
  </div>
</div>
`;

exports[`ChangeOtherList renders when move and handle change when selected 1`] = `
<div>
  <div
    class="form-check-inline mb-3"
  >
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-copy"
        name="copy"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-copy"
        title=""
      >
        Copy
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="move-action-move"
        name="move"
        type="radio"
        value="true"
      />
      <label
        class="form-check-label"
        for="move-action-move"
        title=""
      >
        Move
      </label>
    </div>
  </div>
  <div
    class="mb-3"
  >
    Choose an existing list or create a new one.
  </div>
  <button
    class="mb-3 btn btn-link"
    type="button"
  >
    Create new list
  </button>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="existingList"
    >
      Existing list
    </label>
    <select
      class="form-control"
      id="existingList"
      name="existingList"
    >
      <option
        disabled=""
        value=""
      >
        Select Existing list
      </option>
      <option
        value="1"
      >
        foobar
      </option>
    </select>
  </div>
</div>
`;

exports[`ChangeOtherList shows choose existing list link when existingListsOptions and correct fields when showNewList and copy 1`] = `
<div>
  <div
    class="form-check-inline mb-3"
  >
    <div
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="move-action-copy"
        name="copy"
        type="radio"
        value="true"
      />
      <label
        class="form-check-label"
        for="move-action-copy"
        title=""
      >
        Copy
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-move"
        name="move"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-move"
        title=""
      >
        Move
      </label>
    </div>
  </div>
  <div
    class="mb-3"
  >
    Choose an existing list or create a new one.
  </div>
  <button
    class="mb-3 btn btn-link"
    type="button"
  >
    Choose existing list
  </button>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="newListName"
    >
      New list name
    </label>
    <input
      class="form-control"
      id="newListName"
      name="newListName"
      placeholder="My super cool list"
      type="text"
      value=""
    />
  </div>
  <div
    class="mt-3 mb-3"
  >
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="updateCurrentItems"
        name="updateCurrentItems"
        type="checkbox"
      />
      <label
        class="form-check-label"
        for="updateCurrentItems"
        title=""
      >
        Would you like to also update the current items?
      </label>
    </div>
  </div>
</div>
`;

exports[`ChangeOtherList shows choose existing list link when existingListsOptions and correct fields when showNewList and move 1`] = `
<div>
  <div
    class="form-check-inline mb-3"
  >
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-copy"
        name="copy"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-copy"
        title=""
      >
        Copy
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="move-action-move"
        name="move"
        type="radio"
        value="true"
      />
      <label
        class="form-check-label"
        for="move-action-move"
        title=""
      >
        Move
      </label>
    </div>
  </div>
  <div
    class="mb-3"
  >
    Choose an existing list or create a new one.
  </div>
  <button
    class="mb-3 btn btn-link"
    type="button"
  >
    Choose existing list
  </button>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="newListName"
    >
      New list name
    </label>
    <input
      class="form-control"
      id="newListName"
      name="newListName"
      placeholder="My super cool list"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`ChangeOtherList shows create new list link when not showNewListForm 1`] = `
<div>
  <div
    class="form-check-inline mb-3"
  >
    <div
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="move-action-copy"
        name="copy"
        type="radio"
        value="true"
      />
      <label
        class="form-check-label"
        for="move-action-copy"
        title=""
      >
        Copy
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="move-action-move"
        name="move"
        type="radio"
        value="false"
      />
      <label
        class="form-check-label"
        for="move-action-move"
        title=""
      >
        Move
      </label>
    </div>
  </div>
  <div
    class="mb-3"
  >
    Choose an existing list or create a new one.
  </div>
  <button
    class="mb-3 btn btn-link"
    type="button"
  >
    Create new list
  </button>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="existingList"
    >
      Existing list
    </label>
    <select
      class="form-control"
      id="existingList"
      name="existingList"
    >
      <option
        disabled=""
        value=""
      >
        Select Existing list
      </option>
      <option
        value="1"
      >
        foobar
      </option>
    </select>
  </div>
  <div
    class="mt-3 mb-3"
  >
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="updateCurrentItems"
        name="updateCurrentItems"
        type="checkbox"
      />
      <label
        class="form-check-label"
        for="updateCurrentItems"
        title=""
      >
        Would you like to also update the current items?
      </label>
    </div>
  </div>
</div>
`;
